text-motions.c

```diff
@@ -506,33 +506,40 @@ size_t text_bracket_match(Text *txt, size_t pos) {
 }
 
 static size_t match_symbol(Text *txt, size_t pos, char search, int direction) {
-	char c, current;
+	char c, current, prev_c, next_c;
 	int count = 1;
-	bool instring = false;
+	bool instring = false, inchar = false;
 	Iterator it = text_iterator_get(txt, pos);
 	if (!text_iterator_byte_get(&it, &current))
 		return pos;
+	prev_c = next_c = current;
 	if (direction >= 0) { /* forward search */
 		while (text_iterator_byte_next(&it, &c)) {
-			if (c != current && c == '"')
+			if (!inchar && c != current && c == '"' && prev_c != '\\')
 				instring = !instring;
-			if (!instring) {
+			if (!instring && c != current && c == '\'' && prev_c != '\\')
+				inchar = !inchar;
+			if (!instring && !inchar) {
 				if (c == search && --count == 0)
 					return it.pos;
 				else if (c == current)
 					count++;
 			}
+			prev_c = c;
 		}
 	} else { /* backwards */
 		while (text_iterator_byte_prev(&it, &c)) {
-			if (c != current && c == '"')
+			if (!inchar && c != current && c == '"' && next_c != '\\')
```

Comment:
This looks like the wrong way around?

When you encounter a quote `"` while iterating backwards you would have to check the previous (not next) character for a slash `\`.

More generally, this heuristic based approach will always have some issues ...